### PR TITLE
ListStore changes for more customizability, local items support and more

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR is a combination of the latest changes/experiments I have worked on in the past few weeks. I have combined all of them into a single branch and split the `ListStore` specific part into this PR. That's why the commit count is so high, but the change set is not as scary(?). For the record, I didn't want to squash my commits as I have ran a lot of different experiments and would like to keep them in the history if I ever need to refer back to them.

I think most of the changes in this PR are straightforward. However, some of them are unique to this feature, so I'll try to explain them in some detail.

## ListDescriptor Changes

I felt that `ListDescriptor` was in a good place even before this PR for post pagination. However, it had limited extensibility. Specifically, each list was tied to a local site and adding different filters was always going to be a pain with migrations, nullability and such.

I have been thinking about a different approach since the last hack week and decided to give it ago which I think worked out pretty well. The idea is fairly simple. Instead of saving each field of `ListDescriptor` in DB, we'll only keep 2 values to be able to identify lists.

A `ListModel` is only queried from the DB with a `ListDescriptor` which means there is no other way to access it. That means that we don't actually need to save the fields of `ListDescriptor` in the DB, all we need is a unique identifier. There are several ways to achieve that, but the most common one is to use a hashing algorithm to calculate an integer as the identifier and that's what I have done in this PR. If we want to avoid collisions at all costs, there are perfect hashing algorithms to achieve that. However, for the first iteration, I didn't feel that's necessary, so we are just using the builtin hash calculation.

With this new approach, we can just use an interface to define a `ListDescriptor` and different features can do whatever they need with it. In its simplest form, it'd look like the following:

```
interface ListDescriptor {
    val uniqueIdentifier: Int
}
```

## Type Identifier

A unique identifier would normally be enough for the `ListDescriptor`. However, there are some actions that can affect multiple `ListDescriptor`s at the same time. In some cases, figuring out which `ListDescriptor`s will be affected by a change can be impossible. That's why I have also added a `typeIdentifier` to it.

Consider the following case: The user opens the posts page for `All Posts` filter. They then switch to `Published` posts and take an action that'll change the list (like change the title of a post) then immediately switch back to `All Posts` before the action can be fully completed. In this case, the user would expect the change to be reflected in both `All Posts` and `Published` posts pages. If we were to only update lists based on their unique identifier, we would only update the `Published` posts list. If we use a `typeIdentifier` instead, we can update both lists (and more) at the same time whenever there is a change for that `type`.

The way the `typeIdentifier` will be described will depend on a list. In most cases, there will be a close relationship with how we fetch lists from remote. For posts, the `typeIdentifier` is calculated by using a static string "site-post-list-" and the hash code of the `localSiteId` of the post. So, whenever there is a change for posts of a site, those lists should be updated.

## Other Changes

Other changes I have made are not that exciting and hopefully most of them are properly documented in the code. The changes worth mentioning are:

1. Local items support. The client can provide a list of items to `ListStore` which will be added at the top of the list. A utility function is provided from `ListManager` to make it easier to compare different `ListManager`s since local items make it harder to do that.
2. `LIST_ITEMS_CHANGED` and `LIST_ITEMS_REMOVED` actions are implemented. The change action simply relays the message to the listeners. Remove action will remove the items in the provided list from every list for that `typeIdentifier`. Both of these actions will be triggered from FluxC stores when the underlying data changes for those lists. For example, if a `PostModel` is removed from the DB, the `ListStore` will be notified of that change. (It's NOT a way to manually remove items from lists and I wish I could easily prevent clients from using these actions.)
3. `PostListDescriptor` is implemented which showcases why it's important to make the `ListDescriptor` an interface that's fully customizable. Even for posts, since we have both REST and XML-RPC sites, it gives us an easy way to differentiate their behavior and also have type safety.

Let me know if you have any questions!

/cc @AmandaRiu

### CODE REVIEW

File: fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PostListDescriptor.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    class PostListDescriptorForRestSite(
+        site: SiteModel,
+        val status: PostStatusForRestSite = PostStatusForRestSite.PUBLISH,
+        order: ListOrder = ListOrder.DESC,
+        orderBy: PostListOrderBy = PostListOrderBy.DATE,
+        val searchQuery: String? = null
+    ) : PostListDescriptor(site, orderBy, order) {
+        enum class PostStatusForRestSite(val value: String) {
```
Comment (malinajirka) [inline fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PostListDescriptor.kt]: Hmm, why is the post status tied just to Rest sites? Aren't the same statuses supported on self-hosted sites?

Comment (oguzkocer) [reply fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PostListDescriptor.kt]: Good question. It's because the post statuses can be changed for xml-rpc sites through plugins. So, we can't rely on hard-coded strings for it. We might end up changing both of these descriptors to just accept a string or a different class for status and fetch them from the API. That work is not part of this project (actually none of the filters are), so I just added the statuses for REST sites, for mostly testing and as showcase.

**Comment by oguzkocer** [comment]
Thanks a lot for the review @malinajirka!

> AFAIU local items will always be at the beginning of the list. Is this constraint really necessary? It's ok for current business requirements in posts list, but that might not always be the case. Wdyt?

Unfortunately that constraint is necessary. Since we may not have the remote data, we can't order things in the client side (we also don't want to), so local objects will always be added at the top of the list. I don't consider this to be a shortcoming of this particular solution, but a general constraint. Because even if we had the actual data fetched, if we haven't fetched all of it, mixing remote and local objects and ordering them in the client will move things around as we get more data from the remote. I think, ideally we won't even show local and remote objects in the same list and have filters for them. However, for cases that's not possible or desirable, the component still tries to be helpful while avoiding mentioned issues.

> I've done my best reviewing the code carefully. However, there are a lot of changes and it's difficult to not miss anything. Nevertheless we can always fix it later, if something comes up.

Thank you for your effort. This was an edge case in terms of PR complication and I'll do my best to avoid it in the future. We'll fix whatever comes up in separate PRs.

> The instrumentation tests keep failing on both Travis and local machine. As soon as these are fixed, I think it's ready for merge.

The instrumentation test that's failing is not related to the changes in this PR or this project. I actually reported this issue to @aforcier before, but we couldn't come up with a fix for it or a reliable way to reproduce it. @aforcier Should we remove that test or maybe modify it in some way so as to avoid the random failures?

> (Some classes are missing comments, I assume you are planning to add those later)

I wasn't actually planning on adding more comments to the remaining classes/methods as they are straightforward and following the naming and implementation conventions of FluxC. If you feel any of it requires such documentation, I am happy to add it.

I have been trying to document everything as much as possible where I thought it made sense. However, that can be pretty hard to do and I am sure I either didn't add them where it should be or added too much of it where it wasn't necessary. I am happy to make any changes you see fit because as the reviewer you probably have a better idea of what's useful and what's not.

@malinajirka Hope this all makes sense. Let me know what you think!

**Comment by malinajirka** [comment]
Thanks @oguzkocer  for such an exhaustive answer! ;)

>I wasn't actually planning on adding more comments to the remaining classes/methods as they are straightforward and following the naming and implementation conventions of FluxC. If you feel any of it requires such documentation, I am happy to add it.

I'd consider adding a comment to the `ListDescriptor`. I won't block the PR on it, so if you feel the comment would help, please add it in another/next PR. Thanks!

**Comment by oguzkocer** [comment]
> I'd consider adding a comment to the ListDescriptor. I won't block the PR on it, so if you feel the comment would help, please add it in another/next PR.

🤦‍♂️ I meant to have a rather lengthy documentation on `ListDescriptor`, but somehow I forgot to add that before this PR, you're right, that's a very important interface to document properly.

My next PR should be ready (I'll review in the morning before opening it), but I'll open a separate PR for the documentation with maybe some minor fixes/improvements. Thanks for pointing it out!
